### ISSUE TITLE
b.download() and b.shellExecute() on Windows

### ISSUE THREAD

**Issue by trych** [basiljs/basil.js#135]
Would be nice to have, Windows users in my class are understandably disappointed not to be able to use it.

Does anyone know how to run a Windows shell command via VB-Script? I have absolutely no idea about shell commands on Windows. 😞

**Comment by fabianmoronzirfas** [comment]
Sorry. No idea. The only way I know is creating a .bat file and executing it. But this fires up the cmd.exe. This is not a good solution. Looks like hacker tech taking over your computer.

**Comment by trych** [comment]
> Looks like hacker tech taking over your computer.

Now, this sounds pretty intriguing. 😎

Found [this link](https://forums.adobe.com/thread/783623), where tomaxxi demonstrates how to Window's cmd.exe via a VB-Script. This should be the first step.

Next step is to put together a download shell command for Windows. As `wget` and `curl` are not native to Windows apparently, the alternative seems to be to [use the DownloadFile method of the WebClient object](http://superuser.com/a/362156/591695)*.

\*Full disclaimer: I have absolutely no idea what «the DownloadFile method of the WebClient object» is. 😉

**Comment by fabianmoronzirfas** [comment]
Have fun with VBScript. I wont go down that path.

**Comment by trych** [comment]
Well, I can still open an issue about, can't I? I never said it was the most fun feature to work on, still I think this is a legitimate feature request.

**Comment by ffd8** [comment]
Just had two students in my class last week reach this problem.
I'd place it as a low priority.. but would be nice to find an alternative method...
= a mission for stackoverflow?

How about `b.loadString("http://pathtoapi.com");`? (ie. nytimes.com)
Any luck getting remote API JSON data on windows?

**Comment by fabianmoronzirfas** [comment]
I have an windows env now. I'll take a look

**Comment by fabianmoronzirfas** [comment]
For starters I'll asked in the extendscript slack channel lets see if somebody has a solution. If this does not work I move up to the adobe dev channel and afterwards to sto :-)

This is the post


Dear all. I think this was already discussed several times but I can't find any good example around the web. How do you execute system commands on windows? On macOS we use (in Basil.js) these snippets:  

Snippet 1:  

    var cmd = "pwd";
    app.doScript("return do shell script item 1 of arguments", ScriptLanguage.applescriptLanguage, [cmd]); 

And to make a call to an URL over HTTP/s this one:  

Snippet 2:  

    var url = "https://api.github.com/orgs/basiljs/repos"
    var cmd = "curl -m 15 -L '" + url + "'";
    app.doScript("return do shell script item 1 of arguments", ScriptLanguage.applescriptLanguage, [cmd]);

Last this is our call to download a file from the web.  

Snippet 3 (shell script):  

    #!/bin/bash
    mkdir -p "$1"
    cd "$1"
    if [ -z "$3" ]
      then
        curl -L -O "$2"
      else
        curl -L -o "$3" "$2"
    fi



Snippet 4 (extendscript):  

    var url = 'http://fabianmoronzirfas.me/assets/images/avatar/avatar-0.JPG';
    var cmd = 'sh /Users/icke/tmp/download/download.sh /Users/icke/tmp/download/bin/ ' + url;
    app.doScript('return do shell script item 1 of arguments', ScriptLanguage.applescriptLanguage, [cmd]);


Has anybody an equivalent for snippet 1 + 4 for windows?

**Comment by fabianmoronzirfas** [comment]
I put the essentials into a repo here --> https://github.com/fabianmoronzirfas/extendscript-call-system Also got a response from @grefel (thanks for that) but man VBScript is no fun. 😞  I'll keep you update

**Comment by ffd8** [comment]
Thanks for looking into this! Sure hope it's possible to avoid VBScript??

**Comment by fabianmoronzirfas** [comment]
Nope VBS it is. 😸

**Comment by fabianmoronzirfas** [comment]
This might also be helpful https://msdn.microsoft.com/en-us/library/windows/desktop/aa362813%28v=vs.85%29.aspx

**Comment by grefel** [comment]
As long as you don't want to see a terminal window popping up, you also need VBS to run BITSAdmin Tool

**Comment by trych** [comment]
Just dropping this for later, found some code that seems to execute some script under windows right [here](https://forums.adobe.com/message/9771613#9771613). No idea if it works, but might be worth investigating.

**Comment by trych** [comment]
Ok, good news. Kasyan Servetsky has found a way to execute a VBS download script under Windows (out of the box, no special installations required). Here is the [link to the original discussion](https://forums.adobe.com/message/10079206#10079206).

I have already [asked him for permission](https://twitter.com/askoldich/status/945677561299898368) to use his code (while giving him credit) and he said it's no problem.

Here is Kasyan's code:
```js
function GetDataFromWshShell(webAddress) {    
    var vbs = 'Dim str\r';  
    vbs += 'URL = "http://' + webAddress + '"\r';  
    vbs += 'Set WshShell = CreateObject("WScript.Shell")\r';   
    vbs += 'Set http = CreateObject("Microsoft.XmlHttp")\r';    
    vbs += 'On Error Resume Next\r';    
    vbs += 'http.open "GET", URL, False\r';    
    vbs += 'http.send ""\r';    
    vbs += 'if err.Number = 0 Then\r';    
    vbs += 'str = http.responseText\r';   
    vbs += 'Else\r';    
    vbs += 'MsgBox("error " & Err.Number & ": " & Err.Description)\r';   
    vbs += 'End If\r';    
    vbs += 'set WshShell = Nothing\r';    
    vbs += 'Set http = Nothing\r';   
    vbs += 'Set objInDesign = CreateObject("InDesign.Application")\r';  
    vbs += 'objInDesign.ScriptArgs.SetValue "WshShellResponse", str';  
        
    try {      
        app.doScript(vbs, ScriptLanguage.VISUAL_BASIC, undefined, UndoModes.FAST_ENTIRE_SCRIPT);       
    }    
    catch(err) {    
        alert(err.message + ", line: " + err.line);    
    }   
  
    var str = app.scriptArgs.getValue("WshShellResponse");  
    app.scriptArgs.clear();  
    $.writeln("Output from WshShell:\n" + str);  
}
```

**Comment by b-g** [comment]
Wow! A dream becomes true for all the windows peeps! Super great!

**Comment by fabianmoronzirfas** [comment]
Makes me wonder If we can run the macOS script without writing it to a file.

**Comment by trych** [comment]
> Makes me wonder If we can run the macOS script without writing it to a file.

Yes, absolutely, that's what I was referring to [here](https://github.com/basiljs/basil.js/issues/229#issuecomment-349476424).

**Comment by ffd8** [comment]
@trych - any further ideas/progress in this domain? Again running into justifiably dissapointed students on PC, not being able to download images from APIs... luckily they can handle remote text via `loadStrings()` after #346 – but downloading files still lacks. Not being on a windows machine makes it tricky to do any testing.. @fabianmoronzirfas you mentioned above having a windows environment – any chance to try and build upon this or the loadStrings script?

**Comment by trych** [comment]
Sorry, no progress from me. I was starting to work on this a while back, but then other projects got in the way and now in the lockdown I don't have any access to Windows machines anymore. :/

**Comment by ff6347** [comment]
> you mentioned above having a windows environment – any chance to try and build upon this or the loadStrings script?

Yes I have a Windows computer. I can give it a try

**Comment by ffd8** [comment]
Just had students who wanted to make an API call that required passing an authToken via cURL --header.. = had to use the `shellExecute()` function. When on MacOS = no problem.. but after #346 added Windows support for `loadString()`, should actually extend/move that function to add same Windows capability to `shellExecute()`. Can't test.. but can ask students on Windows to try.

Nevertheless, after adding `user-agent` as an option to the `loadString/loadJSON/loadCSV` as needed by some APIs.. probably makes sense to instead accept an object with things like `{userAgent : String, header : String, custom : Array}` (with 'custom' being ready to add as many additional headers or whatnot as needed in the call.. At the same time, when someone goes beyond simply loading a URL for a GET request to an API, they can probably handle using the `shellExecute()` – just need to get it working on Windows too.. I'll give it a crack.
